cmd/swapd/main.go

```diff
@@ -301,375 +276,153 @@ func runDaemon(c *cli.Context) error {
+		switch {
+		case devXMRTaker:
+			conf.DataDir = defaultXMRTakerDataDir
+		case devXMRMaker:
+			conf.DataDir = defaultXMRMakerDataDir
 		}
 	}
-
-	extendedEC, err := extethclient.NewEthClient(ctx, env, ec, ethPrivKey)
-	if err != nil {
+	if err = common.MakeDir(conf.DataDir); err != nil {
 		return nil, err
 	}
-	// TODO: add configs for different eth testnets + L2 and set gas limit based on those, if not set (#153)
-	extendedEC.SetGasPrice(uint64(c.Uint(flagGasPrice)))
-	extendedEC.SetGasLimit(uint64(c.Uint(flagGasLimit)))
+
+	if c.IsSet(flagBootnodes) {
+		conf.Bootnodes = cliutil.ExpandBootnodes(c.StringSlice(flagBootnodes))
+	}
 
 	deploy := c.Bool(flagDeploy)
 	if deploy {
 		if c.IsSet(flagContractAddress) {
 			return nil, errFlagsMutuallyExclusive(flagDeploy, flagContractAddress)
 		}
-		// Zero out any default contract address in the config, so we deploy
-		cfg.ContractAddress = ethcommon.Address{}
+		// Zero out the contract address, we'll set its final value after deploying
+		conf.SwapFactoryAddress = ethcommon.Address{}
 	} else {
 		contractAddrStr := c.String(flagContractAddress)
 		if contractAddrStr != "" {
 			if !ethcommon.IsHexAddress(contractAddrStr) {
 				return nil, fmt.Errorf("%q is not a valid contract address", contractAddrStr)
 			}
-			cfg.ContractAddress = ethcommon.HexToAddress(contractAddrStr)
+			conf.SwapFactoryAddress = ethcommon.HexToAddress(contractAddrStr)
 		}
 
-		if bytes.Equal(cfg.ContractAddress.Bytes(), ethcommon.Address{}.Bytes()) {
+		if conf.SwapFactoryAddress == (ethcommon.Address{}) {
 			return nil, fmt.Errorf("flag %q or %q is required for env=%s", flagDeploy, flagContractAddress, env)
 		}
 	}
 
+	return conf, nil
+}
+
+// validateOrDeployContracts validates or deploys the swap factory. The SwapFactoryAddress field
+// of envConf should be all zeros if deploying and its value will be replaced by the new deployed
+// contract.
+func validateOrDeployContracts(c *cli.Context, envConf *common.Config, ec extethclient.EthClient) error {
+	deploy := c.Bool(flagDeploy)
+	if deploy {
+		if envConf.SwapFactoryAddress != (ethcommon.Address{}) {
+			panic("contract address should have been zeroed when envConf was initialized")
+		}
+	}
+
 	// forwarderAddress is set only if we're deploying the swap contract, and the --forwarder-address
 	// flag is set. otherwise, if we're deploying and the flag isn't set, we also deploy the forwarder.
 	var forwarderAddress ethcommon.Address
 	forwarderAddressStr := c.String(flagForwarderAddress)
 	if deploy && forwarderAddressStr != "" {
-		ok := ethcommon.IsHexAddress(forwarderAddressStr)
-		if !ok {
-			return nil, errors.New("forwarder-address is invalid")
+		if !ethcommon.IsHexAddress(forwarderAddressStr) {
+			return errors.New("forwarder-address is invalid")
 		}
 
 		forwarderAddress = ethcommon.HexToAddress(forwarderAddressStr)
 	} else if !deploy && forwarderAddressStr != "" {
 		log.Warnf("forwarder-address is unused")
```

Comment:
should this return an error?

Reply:
I'll make it return an error. I actually thought it should return an error, but didn't know if it was written this way intentionally.

Reply:
It returns an error now and I added a unit test.  I'll resolve this one.